system_tests/caff_node_test.go

```diff
@@ -169,6 +176,188 @@ func TestEspressoCaffNode(t *testing.T) {
 	err = checkTransferTxOnL2(t, ctx, builderCaffNode, "User17", builder.L2Info)
 	Require(t, err)
 }
+func TestEspressoCaffNodeDelayedMessagesConfirmations(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	valNodeCleanup := createValidationNode(ctx, t, true)
+	defer valNodeCleanup()
+
+	builder, cleanup := createL1AndL2Node(ctx, t, true, false)
+	defer cleanup()
+
+	err := waitForL1Node(ctx)
+	Require(t, err)
+
+	cleanEspresso := runEspresso()
+	defer cleanEspresso()
+
+	// wait for the builder
+	err = waitForEspressoNode(ctx)
+	Require(t, err)
+
+	newAccount := "User16"
+	l2Info := builder.L2Info
+	l2Info.GenerateAccount(newAccount)
+	addr := l2Info.GetAddress(newAccount)
+
+	// Set caff node config variables
+	builder.nodeConfig.EspressoCaffNode.WaitForConfirmations = true
+	builder.nodeConfig.EspressoCaffNode.RequiredBlockDepth = 6
+	builder.nodeConfig.EspressoCaffNode.WaitForFinalization = false
+
+	// start the node
+	log.Info("Starting the caff node")
+	builder2, cleanupCaffNode := createCaffNode(ctx, t, builder)
+	builderCaffNode := builder2.L2
+	defer cleanupCaffNode()
+
+	// Transfer via the delayed inbox
+	delayedTx := l2Info.PrepareTx("Owner", newAccount, 3e7, transferAmount, nil)
+	log.Info("Delayed tx", "delayedtx", delayedTx)
+	tx := builder.L1.SendWaitTestTransactions(t, []*types.Transaction{
+		WrapL2ForDelayed(t, delayedTx, builder.L1Info, "Faucet", 100000),
+	})
+	// Check the caff node RPC for tx. assert that it is not there.
+	_, _, err = builderCaffNode.Client.TransactionByHash(ctx, tx[0].TxHash)
+	ExpectErr(t, err, ethereum.NotFound)
+	// Wait for the tx header to have required confirmations.
+	err = waitForWith(ctx, 240*time.Second, 1*time.Second, func() bool {
+		header, err := builder.L1.Client.HeaderByNumber(ctx, nil) // get the latest header to check tx block depth
+		Require(t, err)
+		return header.Number.Int64() >= tx[0].BlockNumber.Int64()+int64(builder.nodeConfig.EspressoCaffNode.RequiredBlockDepth) // check that the tx is at least RequiredBlockDepth blocks deep in the parent chains state.
+	})
+	// Wait for the tx to appear on the caff node
+	err = waitForWith(ctx, 240*time.Second, 10*time.Second, func() bool {
```

Comment:
I guess this test seems kind of not testing what we want it to test because it first waits for the required height and then we check that caff node balance updates. 

It should actually test this:
We continuously monitor that Caff node balance doesnt update until the header number has been reached. Only after it has reaches + a few seconds the balance updates.
It should throw an error if Caff node balance updates before the header.Number has not reached

Reply:
Sure, I have a somewhat unorthodox idea with go-routines that will test these concurrently.

Reply:
I think the simple way to do it would be to get the current block number, check if it is less than the header.Number, the caff node has not updated its balance. If it has updated the balance thrown an error, if header.Number is reached wait for a few seconds and see if the balance updates. But if you have a better way of doing it feel free to do so! Just an initial thought I had

Reply:
I see your point, but wouldn't this still have the same problem of "The caff node could update it's balance before the delayed message is finalized" because if we only check the header and balance once, there is nothing asserting that the caff node hasn't at all updated its balance before the delayed message is finalized.

I would argue its a similar pattern to a [TOC/TOU](https://en.wikipedia.org/wiki/Time-of-check_to_time-of-use) attack. Basically we can't guarentee that after we check the first condition, the state of the assertion we just tested won't update before we check the second condition.